homeassistant/components/imap/sensor.py

```diff
@@ -54,139 +46,56 @@ async def async_setup_platform(
-        config.get(CONF_SEARCH),
+    async_create_issue(
+        hass,
+        DOMAIN,
+        "deprecated_yaml",
+        breaks_in_ha_version="2022.11.0",
+        is_fixable=False,
+        severity=IssueSeverity.WARNING,
+        translation_key="deprecated_yaml",
+    )
+    hass.async_create_task(
+        hass.config_entries.flow.async_init(
+            DOMAIN,
+            context={"source": SOURCE_IMPORT},
+            data=config,
+        )
     )
-    if not await sensor.connection():
-        raise PlatformNotReady
 
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, sensor.shutdown)
-    async_add_entities([sensor], True)
 
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Imap sensor."""
 
-class ImapSensor(SensorEntity):
-    """Representation of an IMAP sensor."""
+    coordinator: ImapDataUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
 
-    def __init__(self, name, user, password, server, port, charset, folder, search):
-        """Initialize the sensor."""
-        self._name = name or user
-        self._user = user
-        self._password = password
-        self._server = server
-        self._port = port
-        self._charset = charset
-        self._folder = folder
-        self._email_count = None
-        self._search = search
-        self._connection = None
-        self._does_push = None
-        self._idle_loop_task = None
+    async_add_entities([ImapSensor(coordinator)])
 
-    async def async_added_to_hass(self) -> None:
-        """Handle when an entity is about to be added to Home Assistant."""
-        if not self.should_poll:
-            self._idle_loop_task = self.hass.loop.create_task(self.idle_loop())
 
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return self._name
+class ImapSensor(CoordinatorEntity[ImapDataUpdateCoordinator], SensorEntity):
+    """Representation of an IMAP sensor."""
 
-    @property
-    def icon(self):
-        """Return the icon to use in the frontend."""
-        return ICON
+    _attr_icon = "mdi:email-outline"
+    _attr_has_entity_name = True
 
-    @property
-    def native_value(self):
-        """Return the number of emails found."""
-        return self._email_count
-
-    @property
-    def available(self) -> bool:
-        """Return the availability of the device."""
-        return self._connection is not None
+    def __init__(self, coordinator: ImapDataUpdateCoordinator) -> None:
+        """Initialize the sensor."""
+        super().__init__(coordinator)
+        # To be removed when YAML import is removed
+        if CONF_NAME in coordinator.config_entry.data:
+            self._attr_name = coordinator.config_entry.data[CONF_NAME]
+            self._attr_has_entity_name = False
+        self._attr_unique_id = f"{coordinator.config_entry.entry_id}-{coordinator.config_entry.data[CONF_USERNAME]}"
```

Comment:
I don't think we need to add the username if we're using the config entry id as unique id.

```suggestion
        self._attr_unique_id = coordinator.config_entry.entry_id
```